### PR TITLE
feat: add firework shows

### PR DESCRIPTION
**Pull request by Sqrl34**
## Description

Fireworks show for the 4th. Add multiple locations with multiple scenes with various events all at the same time. Shows started through a command and a permission system.

Known bug: The object does not spawn when doing shows but does spawn with individual fireworks.

## Checklist

- [x] I have personally loaded this code into an updated Qbox project and checked all of its functionality.
- [x] My pull request fits the contribution guidelines & code conventions.

### CODE REVIEW

File: config/shared.lua [hunk 1]
```diff
@@ -49,4 +49,115 @@ return {
             },
         },
     },
+
+    shows = {
+        ['lsia'] = { -- location for command
+            { -- Sequences
```
Comment (Manason) [inline config/shared.lua]: What's the benefit of having difference sequences versus one show being one sequence?

File: client/main.lua [hunk 2]
```diff
@@ -75,3 +105,17 @@ lib.callback.register('qbx_fireworks:client:useFirework', function(asset)
...
+    if show ~= nil then
+        for i=1, #sharedConfig.shows[show] do -- Sequences
+            for j=1, #sharedConfig.shows[show][i].fireworks do -- Individual
+                TriggerServerEvent('qbx_fireworks:server:spawnShowObject', sharedConfig.shows[show][i].fireworks[j].asset, sharedConfig.shows[show][i].fireworks[j].coords)
+                startFireworkShow(sharedConfig.shows[show][i].fireworks[j].asset, sharedConfig.shows[show][i].fireworks[j].coords, sharedConfig.shows[show][i].fireworks[j].height)
+                Wait(sharedConfig.shows[show][i].fireworks[j].wait)
+            end
+            Wait(sharedConfig.shows[show][i].masterWait)
+        end
+    end
```
Comment (Manason) [inline client/main.lua]: using variables will make the code more readable. Avoid referencing the same table value twice.
https://manason.github.io/effective-fivem-lua/tables/#extract-duplicate-table-dereferences-into-local-variables

File: client/main.lua [hunk 3]
```diff
@@ -75,3 +105,17 @@ lib.callback.register('qbx_fireworks:client:useFirework', function(asset)
         return false
     end
 end)
+
+RegisterNetEvent('qbx_fireworks:client:startShow', function(show)
+    if show ~= nil then
```
Comment (Manason) [inline client/main.lua]: https://manason.github.io/effective-fivem-lua/functions/#use-guard-clauses

**Review by Manason — CHANGES_REQUESTED** [review]

File: config/shared.lua (hunk 1 above)
Comment (Sqrl34) [reply config/shared.lua]: The benefit of having different sequences just for visual separation between different segments in a show. If you really wanted to you could just have 1 sequence with a bunch of events with a longer wait time but with sequences, I feel it is easier to divide the show into different blocks.

**Comment by Sqrl34** [comment]
added 2 new locales to the en.json so that will have to be updated in the other languages to

File: config/shared.lua (hunk 1 above)
Comment (Manason) [reply config/shared.lua]: Couldn't the same thing be accomplished with comments? I would get rid of sequences as a data type in favor of a more simple representation of the data

**Review by Manason — CHANGES_REQUESTED** [review]
Maybe the show should be server based rather than client based? What happens if the starting client disconnects in the middle of it? What is responsible for showing fireworks on other clients? Right now it looks as if only the starting client would see fireworks

**Comment by Sqrl34** [comment]
Tests Ran from the above revision
1. Player 1 in paleto and Player 2 at lsia -> player 2 sees/hears fireworks and player 1 hears faint explosion

2. Player 1 starts fireworks show and Player 2 remains in game -> show continues to go on regardless of show initiator leaving

3. Player 1 starts show and Player 2 joins server while show is in progress -> Player 2 hears/sees current show happening

4. Player 1 starts show and teleports away while player 2 remains -> show continues as normal

File: server/main.lua [hunk 4]
```diff
@@ -44,9 +44,20 @@ end)
 lib.addCommand('startshow', {
```
Comment (Manason) [inline server/main.lua:63]: Should we expose this function as an export as well?

File: server/main.lua [hunk 5]
```diff
@@ -44,9 +44,20 @@ end)
...
-        {name = "Location", help = locale('start_show_param_hint'), type = 'string'}
+        { name = "Location", help = locale('start_show_param_hint'), type = 'string' }
     },
     restricted = 'group.admin'
 }, function(source, args, raw)
-    TriggerClientEvent('qbx_fireworks:client:startShow', source, args.Location)
+    local show = args.Location
+    if not show then return end
+    if not config.shows[show] then return end
+    for i = 1, #config.shows[show].fireworks do -- Sequences
```
Comment (Manason) [inline server/main.lua]: Consider extracting config.shows[show] to a variable.

File: server/main.lua [hunk 6]
```diff
@@ -44,9 +44,20 @@ end)
...
+        local firework = config.shows[show].fireworks[i]
+        local fireworkAsset = firework.asset
+        local fireworkCoords = firework.coords
+        local fireworkHeight = firework.height
+        local fireworkWait = firework.wait
```
Comment (Manason) [inline server/main.lua]: Not sure there is any readability gain by storing these variables

**Review by Manason — CHANGES_REQUESTED** [review]
Could more of the client show logic be moved to the server? This is still setup so than a single client still owns the show.

What happens if a client connects in the middle of a show?

**Comment by Sqrl34** [comment]
Regardless of who starts the show, where they are in the world, or if a player or the host leaves or joins the server, all the fireworks will be synced among all clients now.

File: config/shared.lua (hunk 1 above)
Comment (Sqrl34) [reply config/shared.lua]: sequences have been removed, to achieve same effect just increase a wait time

File: server/main.lua (hunk 6 above)
Comment (Sqrl34) [reply server/main.lua]: removed variables

File: server/main.lua (hunk 5 above)
Comment (Sqrl34) [reply server/main.lua]: outdated

File: server/main.lua (hunk 4 above)
Comment (Sqrl34) [reply server/main.lua:63]: possibly, added export

File: client/main.lua (hunk 3 above)
Comment (Sqrl34) [reply client/main.lua]: outdated

File: client/main.lua (hunk 2 above)
Comment (Sqrl34) [reply client/main.lua]: outdated

**Comment by Sqrl34** [comment]
The firework cake object now spawns correctly. All should be good _hopefully_

**Review by Manason — APPROVED** [review]

**Review by KostaZx — APPROVED** [review]

**Pull request merged**